doc/Moving-to-single-app-project-structure.md

```diff
@@ -0,0 +1,160 @@
+
+### Includes
+
+This is the big one. Since we don't use the umbrella structure anymore, all `-include(..)`
+directives pointing to `ejabberd/include` need to be rewritten.
+
+(SED is `sed` on Linuxes or `gsed` on Macs)
+
+* `gsed -i 's/-include_lib("ejabberd\/include\//-include("/' -- src/*.erl src/*.hrl include/*.hrl test/*.erl`
+* `gsed -i 's/-include_lib("ejabberd\/src\//-include("/' -- src/*.erl src/*.hrl include/*.hrl test/*.erl`
+
+### Release scripts
+
+In `rel/files/mongooseim` the variable `EJABBERD_SO_PATH` contains `ejabberd-2.1.8*` fragment.
+Replace it with `mongooseim-*`.
+
+In `rel/files/mongooseimctl` the variable `EJABBERD_EBIN_PATH` contains `ejabberd-*` fragment.
+Replace it with `mongooseim-*`.
+
+### Tests & tools
+
+After previous changes, lots of tests fail, mostly because we changed the application name from
+`ejabberd` to `mongooseim`. Functions to look for are:
+- `application:set_env`
+- `application:get_env`
+- `application:get_key`
+- `application:load`
+- `application:start`
+- `application:stop`
+- `application:ensure_all_started`
+- `code:priv_dir`
+
+Those function are also called via RPC from big test suites. A good method to find them all is to
+search using `application.*ej` regex.
+
+Files which require changes:
+- `include/ejabberd.hrl`
+- `src/ejabberd.erl`
+- `src/ejabberd_app.erl`
+- `src/ejabberd_config.erl`
+- `src/mod_version.erl`
+- `src/translate.erl`
+- `test/ejabberd_config_SUITE.erl`
+- `test/ejabberd_helper.erl`
+- `test/ejabberd_listener_SUITE.erl`
+- `test/zlib_driver_SUITE.erl`
+- `test.disabled/ejabberd_tests/README.md`
+- `test.disabled/ejabberd_tests/run_common_test.erl`
+- `test.disabled/ejabberd_tests/tests/conf_reload_SUITE.erl`
+- `test.disabled/ejabberd_tests/tests/connect_SUITE.erl`
+- `test.disabled/ejabberd_tests/tests/reload_helper.erl`
+- `tools/extract_translations/extract_translations.erl`
+- `tools/xep_tool/xep_tool.escript`
+- `src/mongoose_cluster.erl` (2 calls to `when_app_stopped/1` function)
+
+We also need to remove remaining occurences of `apps/ejabberd` directory structure. Some of the
+files are scripts assembling the path programatically - the trick is to search for the `apps/` keyword.
+
+Affected files:
+- `CONTRIBUTING.md`
+- `Makefile`
+- `elvis.config`
+- `src/amp_strategy.erl`
+- `src/mod_mam_odbc_arch.erl`
+- `test.disabled/ejabberd_tests/Makefile`
+- `test.disabled/ejabberd_tests/run_common_test.erl`
+- `test.disabled/ejabberd_tests/tests/vcard_SUITE.erl`
+- `tools/cd_tools/provision_mysql.sh`
+- `tools/cd_tools/provision_pgsql.sh`
+- `tools/fill_roster.erl`
+- `tools/privacy.erl`
+- `tools/travis-setup-db.sh`
+- `tools/travis-test.sh`
+- `tools/travis-upload-to-s3.sh`
+- `tools/zabbix_add_counters.py`
+
+## Moving a MongooseIM's fork to a new structure
+
+This section describes how to move customized fork of MongooseIM to a new structure.
+
+1. Move all directories from `apps/ejabberd` directory to the repository root, e.g. `apps/ejabberd/src/`
```

Comment:
Move all directories from the `apps/ejabberd` directory to the repository root, e.g. `apps/ejabberd/src/`